File: LC 857.cpp
class Solution {
public:
    double mincostToHireWorkers(vector<int>& quality, vector<int>& wage, int k) {
        double ans = 0x7f7f7f7f;
        int n = quality.size();
        vector<pair<double, int>> workers;
        for (int i = 0; i < n; ++i) {
            workers.emplace_back(1.0 * wage[i] / quality[i], quality[i]);
        }
        sort(workers.begin(), workers.end());
        priority_queue<int> pq;
        int sum = 0;
        for (int i = 0; i < n; i++) {
            int t = workers[i].second;
            sum += t;
            pq.push(t);
            while (pq.size() > k) {
                sum -= pq.top();
                pq.pop();
            }
            if (pq.size() == k) {
                ans = min(ans, sum * workers[i].first);
            }
        }
        return ans;
    }
};

File: 1000/LC 857.go
func mincostToHireWorkers(quality []int, wage []int, k int) float64 {
    n := len(quality)
    var ans float64 = 0x7f7f7f7f
    worker := make([]float64, n)
    pos := make([]int, n)
    for i := range quality {
        pos[i] = i
        worker[i] = float64(wage[i]) / float64(quality[i])
    }
    sort.Slice(pos, func(i, j int) bool {
        return worker[pos[i]] < worker[pos[j]]
    })

    h := hp{}
    heap.Init(&h)
    sum := 0
    for _, i := range pos {
        q, w := quality[i], worker[i]
        sum += q
        heap.Push(&h, q)
        if h.Len() > k {
            t := heap.Pop(&h).(int)
            sum -= t
        }
        if h.Len() == k {
            ans = min(ans, float64(sum) * w)
        }
    }

    return ans 
}

type hp []int
func (h hp) Len() int {return len(h)}
func (h hp) Less(i, j int) bool {return h[i] > h[j]}
func (h hp) Swap(i, j int) {h[i], h[j] = h[j], h[i]}
func (h *hp) Push(x any) {
    *h = append(*h, x.(int))
}
func (h *hp) Pop() any {
    old := *h
    n := len(old)
    x := old[n-1]
    *h = old[:n-1]
    return x
}